homeassistant/components/obihai/sensor.py

```diff
@@ -23,52 +25,113 @@ async def async_setup_entry(
 
     requester: ObihaiConnection = hass.data[DOMAIN][entry.entry_id]
 
-    sensors = []
+    sensors: list[ObihaiEntity] = []
     for key in requester.services:
-        sensors.append(ObihaiServiceSensors(requester, key))
+        sensors.append(ObihaiServiceSensor(requester, key))
 
     if requester.line_services is not None:
         for key in requester.line_services:
-            sensors.append(ObihaiServiceSensors(requester, key))
+            sensors.append(ObihaiLineServiceSensor(requester, key))
 
     for key in requester.call_direction:
-        sensors.append(ObihaiServiceSensors(requester, key))
+        sensors.append(ObihaiCallDirectionSensor(requester, key))
 
     async_add_entities(sensors, update_before_add=True)
 
 
-class ObihaiServiceSensors(SensorEntity):
-    """Get the status of each Obihai Lines."""
+class ObihaiSensor(ObihaiEntity, SensorEntity, ABC):
```

Comment:
I think this PR should make proper use of Entity descriptions instead of constructing all these classes.
You could set name, icon, service etc. in their and it would cleanup this pretty much.

Reply:
Doesn't seem to me like that would simplify this PR in any significant way.  Dataclasses can't have definitions for the unique update functions that each of these sensor types needs, so the existing classes would have to continue to exist.  In some cases we may be able to get away with handing these sensor functions entityDescriptions, but as the icons are constant (EG: icon changes is phone is off-hook) ... I'm not sure dataclasses would be able to handle the vast majority of what's happening here.

I'll sleep on it and take another look tomorrow, but I doubt there's more than one or two places where dataclasses would even work.

Reply:
I think all off it can be covered by it.
Example here but you can store references to a callable etc. etc. so all the definitions are saved in the entity_description for each entity and more or less you would only need one class to cover all needs.

Reply:
Sorry, forgot the link. Edited by previous comment so now there is a link.
Notice that entity descriptions can be extended and more or less contain whatever to simplify the code further down.
My example is a pretty easy and straight forward one.

Reply:
I've converted this to draft, with an initial attempt at doing what you've suggested, there are still some issues, and it needs a bit of cleanup.  So far I'm not convinced this reduces complexity...